packages/store-ui/src/molecules/NumberInput/NumberInput.tsx

```diff
@@ -0,0 +1,160 @@
+import React, { useCallback, useEffect, useState } from 'react'
+import type { PropsWithChildren } from 'react'
+
+import Button from '../../atoms/Button'
+import Input from '../../atoms/Input'
+
+interface ActionButtonProps {
+  actionOption: 'dec' | 'inc'
+  disabled: boolean
+  ariaLabel: string
+  onClick: (currentValue: number) => void
+}
+
+const ActionButton = ({
+  children,
+  actionOption,
+  disabled,
+  ariaLabel,
+  onClick,
+}: PropsWithChildren<ActionButtonProps>) => (
+  <Button
+    type="button"
+    data-store-button={actionOption}
+    onClick={(event) => onClick(parseInt(event.currentTarget.value, 10))}
+    disabled={disabled}
+    aria-label={ariaLabel}
+    aria-disabled={disabled}
+  >
+    {children}
+  </Button>
+)
+
+export interface NumberInputProps {
+  /**
+   * The initial value of the input. Should be less than max and greater then min.
+   */
+  defaultValue?: number
+  /**
+   * The minimum value of the input. When the input value reaches the minimum, the decrease's button is disabled.
+   */
+  min?: number
+  /**
+   * The maximum value of the input. When the input value reaches the maximum, the increase's button is disabled.
+   */
+  max?: number
+  /**
+   * Define if the component is disabled.
+   */
+  disabled?: boolean
+  /**
+   * Callback that fires when the current value changes, either by directly changing the input or by buttons' action.
+   */
+  onChange?: (value: number) => void
+  /**
+   * ID to find this component in testing tools (e.g.: cypress, testing library, and jest).
+   *
+   * @default 'store-number-input'
+   */
+  testId?: string
+}
+
+const NumberInput = ({
+  defaultValue,
+  min = -Infinity,
+  max = Infinity,
+  disabled = false,
+  testId = 'store-number-input',
+  onChange,
+}: NumberInputProps) => {
+  const initialValue =
+    defaultValue && defaultValue >= min && defaultValue <= max
+      ? defaultValue
+      : min !== -Infinity
+      ? min
+      : 0
+
+  const [currentValue, setCurrentValue] = useState(initialValue)
+  const [buttonDisabled, setButtonDisabled] = useState<{
+    dec: boolean
+    inc: boolean
+  }>()
+
+  const handleCurrentValueUpdate = useCallback(
```

Comment:
If I understand well, the usage of the useCallback isn't necessary and harm performance. Take a look on this article https://kentcdodds.com/blog/usememo-and-usecallback